lib/HLSL/DxilCondenseResources.cpp

```diff
@@ -2781,12 +2788,409 @@ ModulePass *llvm::createDxilAllocateResourcesForLibPass() {
+          // Set incoming values to undef until next pass
+          newPhi->addIncoming(UndefValue, Phi->getIncomingBlock(j));
+        }
+        mergeHeapArgs(Phi, newPhi, Phi->incoming_values());
+      } else if (SelectInst *Sel = dyn_cast<SelectInst>(Select)) {
+        IRBuilder<> B(Sel);
+        Value *newSel =
+            B.CreateSelect(Sel->getCondition(), UndefValue, UndefValue);
+        User::op_range range = User::op_range(Sel->getOperandList() + 1,
+                                       Sel->getOperandList() + 3);
+        mergeHeapArgs(
+            Sel, newSel,
+            range);
+      } else {
+        DXASSERT(false, "otherwise, non-select/phi in Selects set");
+      }
+    }
+  }
+  // propagate CreateHandleFromHeapArgs for HandleSel which all operands are
+  // other HandleSel.
+  void PropagateHeapArgs() {
+    SmallVector<Value *, 4> Candidates;
+    for (auto &Select : HandleSelects) {
+      CreateHandleFromHeapArgs &args = HandleToArgs[Select];
+      if (args.isResolved)
+        continue;
+      Candidates.emplace_back(Select);
+    }
+
+    while (1) {
+      SmallVector<Value *, 4> Candidates2;
+      for (auto &Select : Candidates) {
+        CreateHandleFromHeapArgs &args = HandleToArgs[Select];
+        if (PHINode *Phi = dyn_cast<PHINode>(Select)) {
+          mergeHeapArgs(Phi, args.Index, Phi->incoming_values());
+        } else if (SelectInst *Sel = dyn_cast<SelectInst>(Select)) {
+          User::op_range range = User::op_range(Sel->getOperandList() + 1,
+                                                Sel->getOperandList() + 3);
+          mergeHeapArgs(
+              Sel, args.Index,
+              range);
+        } else {
+          DXASSERT(false, "otherwise, non-select/phi in Selects set");
+        }
+        // Reload args incase it is changed in mergeHeapArgs.
+        args = HandleToArgs[Select];
+        if (args.isResolved)
+          continue;
+        Candidates2.emplace_back(Select);
+      }
+      // Some node cannot be reached.
+      if (Candidates2.size() == Candidates.size())
+        return;
+      Candidates = Candidates2;
+    }
+  }
+
+  void UpdateSelectsForHandleSelect(hlsl::OP *hlslOP) {
+    if (HandleSelects.empty())
+      return;
+    LLVMContext &Ctx = HandleSelects[0]->getContext();
+    Type *pVoidTy = Type::getVoidTy(Ctx);
+    // NOTE: currently phi of createHandleFromHeap and createHandleFromBinding
+    // is not supported.
+    Function *createHdlFromHeap =
+        hlslOP->GetOpFunc(DXIL::OpCode::CreateHandleFromHeap, pVoidTy);
+    Value *hdlFromHeapOP = hlslOP->GetI32Const(
+        static_cast<unsigned>(DXIL::OpCode::CreateHandleFromHeap));
+    for (auto &Select : HandleSelects) {
+      if (PHINode *Phi = dyn_cast<PHINode>(Select)) {
+        unsigned numIncoming = Phi->getNumIncomingValues();
+        CreateHandleFromHeapArgs &args = HandleToArgs[Phi];
+        PHINode *newPhi = cast<PHINode>(args.Index);
+        bool bAllIndexResolved = true;
```

Comment:
All checking related to `bAllIndexResolved` and additional merging seems like it should be unnecessary now that isResolved is on the args for this phi.  If not resolved, we can't replace it.  If it is resolved, we can, and all incoming args with values must be there.  Right?